fairlearn/datasets/_fetch_acsincome.py

```diff
@@ -0,0 +1,158 @@
+
+    Returns
+    -------
+    dataset : :obj:`~sklearn.utils.Bunch`
+        Dictionary-like object, with the following attributes.
+
+        data : ndarray, shape (1664500, 10)
+            Each row corresponding to the 10 feature values in order.
+            If ``as_frame`` is True, ``data`` is a pandas object.
+        target : numpy array of shape (1664500,)
+            Integer denoting each person's annual income.
+            A threshold can be applied as a postprocessing step to frame this
+            as a binary classification problem.
+            If ``as_frame`` is True, ``target`` is a pandas object.
+        feature_names : list of length 10
+            Array of ordered feature names used in the dataset.
+        DESCR : string
+            Description of the ACSIncome dataset.
+
+    (data, target) : tuple of (numpy.ndarray, numpy.ndarray)
+        if ``return_X_y`` is True and ``as_frame`` is False
+
+    (data, target) : tuple of (pandas.DataFrame, pandas.Series)
+        if ``return_X_y`` is True and ``as_frame`` is True
+
+    References
+    ----------
+    .. [1] Ding, F., Hardt, M., Miller, J., & Schmidt, L. (2021).
+       "Retiring Adult: New Datasets for Fair Machine Learning."
+       Advances in Neural Information Processing Systems, 34.
+
+    """
+    # State Code based on 2010 Census definitions
+    _STATE_CODES = {'AL': '01', 'AK': '02', 'AZ': '04', 'AR': '05', 'CA': '06',
+                    'CO': '08', 'CT': '09', 'DE': '10', 'FL': '12', 'GA': '13',
+                    'HI': '15', 'ID': '16', 'IL': '17', 'IN': '18', 'IA': '19',
+                    'KS': '20', 'KY': '21', 'LA': '22', 'ME': '23', 'MD': '24',
+                    'MA': '25', 'MI': '26', 'MN': '27', 'MS': '28', 'MO': '29',
+                    'MT': '30', 'NE': '31', 'NV': '32', 'NH': '33', 'NJ': '34',
+                    'NM': '35', 'NY': '36', 'NC': '37', 'ND': '38', 'OH': '39',
+                    'OK': '40', 'OR': '41', 'PA': '42', 'RI': '44', 'SC': '45',
+                    'SD': '46', 'TN': '47', 'TX': '48', 'UT': '49', 'VT': '50',
+                    'VA': '51', 'WA': '53', 'WV': '54', 'WI': '55', 'WY': '56',
+                    'PR': '72'}
+    # number of features
+    _NUM_FEATS = 10
+
+    # check that user-provided state abbreviations are valid
+    if states is not None:
+        states = [state.upper() for state in states]
+        for state in states:
+            try:
+                _STATE_CODES[state]
+            except KeyError:
+                raise KeyError(f'Error with state code: {state}\n'
+                               f'State code must be a two letter abbreviation'
+                               f'from the list {list(_STATE_CODES.keys())}\n'
+                               f'Note that PR is the abbreviation for Puerto Rico.'
+                               )
+    else:
+        states = _STATE_CODES.keys()
```

Comment:
Not sure how much we care about performance, but your code will go through the whole dataset and re-add all the rows in this case, which will result in superfluous work given that `fetch_openml` retrieves the whole dataset

Reply:
Correct, it's technically inefficient if users will frequently request data for all states. However, in the other extreme, it should be fast if users frequently request data for just 1 or 2 states. I believe the later will be more common for a couple of reasons:

1) Smaller datasets lead to faster training times, which is convenient when one desires to focus on evaluating fairness rather than dealing with slow training times caused by ~1.7M datapoints. For example, this basically eliminates the possibility of ever training and evaluating an SVC due to terrible scaling with dataset size. Using a linear kernel helps, but practically it's still orders of magnitude slower than decision trees or logistic regression, which would only be compounded if using Exponentiated Gradient to train multiple models.

2) More data isn't necessary to properly evaluate fairness. As the authors conclude: `Increased dataset size does not necessarily help in reducing observed disparities`, so I think requesting data for just 1 or 2 states will more common.

Let me know your thoughts :)

Reply:
If it's a huge issue we can always look into uploading the smaller datasets separately at a later point. I wouldn't worry too much about it as long as it doesn't take too long for you to run locally. Any rough estimate?

Reply:
Yea the rate limiting step is downloading the data from OpenML, which takes about 16 seconds for me. The remaining processing takes well under a second.

Reply:
One suggestion is to use [Pandas's `query` function](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html) when `states is not None`, although I imagine the bigger benefit here is code readability rather than performance, since it's not taking too long for you to run anyway.